New Feature - Responsive-Top-Bar for editor device mode [ED-1161]

joshmarom opened the pull request:
![Screen Shot 2021-02-14 at 21 03 10](https://user-images.githubusercontent.com/2726148/107946235-6d568c00-6f99-11eb-999b-57da6045106f.png)
![Screen Shot 2021-02-14 at 21 03 23](https://user-images.githubusercontent.com/2726148/107946245-6fb8e600-6f99-11eb-81a4-8de55d955811.png)
![Screen Shot 2021-02-14 at 21 02 16](https://user-images.githubusercontent.com/2726148/107946250-70ea1300-6f99-11eb-8d57-ef93eb32130a.png)

joshmarom committed:
Add responsive top bar Marionette template

joshmarom committed:
Merge remote-tracking branch 'Elementor/release/3.1.0' into feature/breakpoints

joshmarom committed:
Merge remote-tracking branch 'Elementor/release/3.1.0' into feature/breakpoints

joshmarom committed:
Progress with top bar UI

joshmarom committed:
Merge remote-tracking branch 'Elementor/release/3.1.0' into feature/breakpoints

joshmarom committed:
Top Bar handling JS switching device mode

joshmarom committed:
CSS for the top-bar and other details in the editor layout

joshmarom committed:
Update the location in which the bar is located

joshmarom committed:
Move the "Device-Mode" from the panel footer to the top-bar

joshmarom committed:
Responsive Top Bar Update

joshmarom committed:
Added functions for entering and exiting device mode

joshmarom committed:
Remove obsolete code

joshmarom committed:
Better naming

joshmarom committed:
Introduce CSS state management properties

joshmarom committed:
Use the new CSS properties to use and set layout state

joshmarom committed:
Add top-bar to resize handling for clean resize and toggle transitions

joshmarom committed:
Handle top-bar events and interactions

joshmarom committed:
Adjust style to compliment changes

joshmarom committed:
Add top-bar to animation definitions

joshmarom committed:
Merge remote-tracking branch 'Elementor/release/3.1.0' into feature/breakpoints

joshmarom committed:
Fix top-bar registration on the editor level

joshmarom committed:
Remove duplicate 'events' class method

rodolphebertozzo commented on includes/editor-templates/responsive-top-bar.php:
```diff
@@ -0,0 +1,79 @@
+
+//$breakpoints = Responsive::get_breakpoints();
+$breakpoints = [
+		'mobile' => __( 'Mobile', 'elementor' ),
+		'tablet' => __( 'Tablet', 'elementor' ),
+		'desktop' => __( 'Desktop', 'elementor' ),
```
```suggestion
	'mobile' => __( 'Mobile', 'elementor' ),
	'tablet' => __( 'Tablet', 'elementor' ),
	'desktop' => __( 'Desktop', 'elementor' ),
```

joshmarom committed:
PHP Lint

joshmarom committed:
Improve open-breakpoints-settings function 

Handle 2 exceptions:
- When editor is in preview mode
- When panel already open on the layout-settings

joshmarom committed:
Remove redundant code

joshmarom committed:
Merge remote-tracking branch 'Elementor/release/3.1.0' into feature/breakpoints

kobizz commented on assets/dev/scss/editor/_responsive-top-bar.scss:
```diff
@@ -0,0 +1,146 @@
+	top: calc(-40px + ( 40px * var(--e-is-device-mode)));
+}
+
+.e-mq { /* Just a prefix. There is no element with this className */
```
Let's discuss this file

joshmarom replied:
Can you be more specific?

kobizz commented on assets/dev/js/editor/top-bar.js:
```diff
@@ -0,0 +1,109 @@
+		elementor.exitDeviceMode();
+	}
+
+	onChange() {
+		const changeCallback = this.getOption( 'change' );
+
+		if ( changeCallback ) {
+			changeCallback();
+		}
+	}
+
+	onClick() {
+		const clickCallback = this.getOption( 'click' );
+
+		if ( clickCallback ) {
+			clickCallback();
+		}
+	}
```
Is it in use?

kobizz commented on assets/dev/js/editor/top-bar.js:
```diff
@@ -0,0 +1,109 @@
+
+	templateHelpers() {
+		return {
+			title: this.getOption( 'title' ),
+			width: '',
+			height: '',
+			activeDevice: '',
```
Are these parameters in use?

joshmarom committed:
Merge remote-tracking branch 'Elementor/release/3.2.0' into feature/breakpoints

joshmarom commented on includes/editor-templates/responsive-top-bar.php:
```diff
@@ -0,0 +1,79 @@
+	exit; // Exit if accessed directly.
+}
+
+//$breakpoints = Responsive::get_breakpoints();
+$breakpoints = [
+	'mobile' => __( 'Mobile', 'elementor' ),
+	'tablet' => __( 'Tablet', 'elementor' ),
+	'desktop' => __( 'Desktop', 'elementor' ),
+];
```
But it returns irrelevant values for my use case.
I left it there because I am aware the current solution isn't right for the next phases. and we need to address those needs, but for now I think it's ok...

joshmarom replied:
Will be improved once responsive settings API is ready

joshmarom commented on assets/dev/js/editor/regions/panel/panel.js:
```diff
@@ -30,34 +30,31 @@ module.exports = BaseRegion.extend( {
 			},
 			resize: function( event, ui ) {
-				elementor.$previewWrapper.css( side, ui.size.width );
+				elementorCommon.elements.$body.css( '--e-editor-panel-width', ui.size.width + 'px' );
+				self.$el.css( 'width', '' );
```
Not sure what the question is

joshmarom commented on assets/dev/js/editor/regions/panel/panel.js:
```diff
@@ -30,34 +30,31 @@ module.exports = BaseRegion.extend( {
 				elementor.getPanelView().updateScrollbar();
 
-				self.saveSize();
+				self.saveStorage( 'size', ui.size.width + 'px' );
```
Not sure what the question is

kobizz replied:
1. Keep using `saveSize` method
2. Change the navigator width to use a custom property

joshmarom commented on assets/dev/js/editor/editor-base.js:
```diff
@@ -640,6 +649,19 @@ export default class EditorBase extends Marionette.Application {
 		}
 	}
 
+	enterDeviceMode() {
+		elementorCommon.elements.$body.addClass( 'e-is-device-mode' );
+	}
+
+	toggleDeviceMode() {
+		elementorCommon.elements.$body.toggleClass( 'e-is-device-mode' );
+	}
+
+	exitDeviceMode() {
+		elementorCommon.elements.$body.removeClass( 'e-is-device-mode' );
+		elementor.changeDeviceMode( 'desktop' );
+	}
+
```
I am not sure this is a feature of the TopBar...
I think it is a state of the entire editor... 

Should we keep it for now, and add a task about it in Jira for the next phase + discussion with @matipojo ?

joshmarom commented on assets/dev/js/editor/regions/panel/footer.js:
```diff
@@ -117,23 +112,8 @@ module.exports = Marionette.ItemView.extend( {
-
-		elementor.changeDeviceMode( newDeviceMode );
+	onDeviceModeIconClick: function() {
+		elementor.toggleDeviceMode();
```
The command wrapper should probably be on the functions above... right?
https://github.com/elementor/elementor/pull/13625#discussion_r567407263

joshmarom committed:
Create Responsive Bar region

joshmarom committed:
Refactor Responsive Bar template

joshmarom committed:
Add Template Placeholder in the editor markup

joshmarom committed:
Init ResponsiveBar when editor ui initiates

joshmarom committed:
Refactoring

joshmarom commented on assets/dev/js/editor/editor-base.js:
```diff
@@ -398,6 +399,14 @@ export default class EditorBase extends Marionette.Application {
 			this.$previewResponsiveWrapper.append( this.$preview );
 		}
 
+		const topBar = new ResponsiveTopBar();
```
Done

kobizz commented on includes/editor-templates/panel.php:
```diff
@@ -78,30 +78,11 @@
 		<span class="elementor-screen-only"><?php echo __( 'History', 'elementor' ); ?></span>
 	</div>
 	<div id="elementor-panel-footer-responsive" class="elementor-panel-footer-tool elementor-toggle-state">
+		<!-- TODO: Change the icon to the new design once the css className has been determined -->
```
`eicon-device-responsive`

kobizz commented on assets/dev/scss/helpers/variables.scss:
```diff
@@ -144,7 +144,7 @@ $editor-red: #d72b3f;
 $editor-warning: #b01b1b;
 $editor-success: #39b54a;
-$editor-info: #71d7f7;
+$editor-info: #58D0F5;
```
Cancel

kobizz commented on assets/dev/scss/editor/_responsive-top-bar.scss:
```diff
@@ -0,0 +1,146 @@
+					font-size: 15px;
+				}
+
+				span {
```
Use tipsy

kobizz commented on assets/dev/scss/editor/_responsive-top-bar.scss:
```diff
@@ -0,0 +1,146 @@
+			padding: 0 !important;
+			position: absolute;
+
+			+ label {
```
Use class

kobizz commented on assets/dev/scss/editor/_responsive-top-bar.scss:
```diff
@@ -0,0 +1,146 @@
+			padding: 4px;
+		}
+
+		button {
```
Use class

joshmarom committed:
Move Init Responsive Bar to after Panel Init

joshmarom committed:
Add panel resize listeners and handlers to top bar region

joshmarom committed:
Remove side-effect from panel resize handler

joshmarom committed:
Add Tipsy to switcher buttons and remove redundant code

joshmarom committed:
Add option Label Ui element

joshmarom committed:
Remove redundant event listener

joshmarom committed:
Remove CSS Tooltip

joshmarom committed:
Minor refactoring of device-switcher markup

joshmarom committed:
Remove side-effect

joshmarom committed:
Use saveSize() function

joshmarom committed:
Hide Label text for non-screen-readers

joshmarom committed:
Revert Color change TBD

joshmarom committed:
Use the new responsive devices Icon

joshmarom committed:
Fix Tipsy

joshmarom committed:
Fix Tipsy

joshmarom committed:
CSS Better selector

joshmarom committed:
Remove redundant templateHelpers() function

joshmarom committed:
Improve CSS selectors

joshmarom committed:
Coding standard

joshmarom committed:
Match template with new css selectors.

matipojo commented on assets/dev/js/editor/regions/responsive-bar/view.js:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	className() {
+		return 'e-mq-bar';
```
`mq`? why not a full name?

joshmarom replied:
For the same reason [other abbreviations](https://www.abbreviations.com/acronyms/WEBSITES) exist.

If you think it should be changed to something else, just tell me what name it should be and I will refactor.

matipojo replied:
The feature is called Responsive Bar. so `rb` will be more understood.

kobizz replied:
The feature name will be `responsive-bar`

matipojo commented on assets/dev/js/editor/regions/responsive-bar/view.js:
```diff
@@ -0,0 +1,100 @@
+		const prefix = '.e-mq-';
+
+		return {
+			switcherOption: prefix + 'switcher__option',
```
Why not hardcoded?

joshmarom replied:
So It can easily be refactored when @matipojo reviews it.

matipojo replied:
Anyway, you need to perform a search-replace on the whole code.

matipojo commented on includes/editor-templates/responsive-top-bar.php:
```diff
@@ -0,0 +1,48 @@
+		</div>
+		<div class="e-mq-bar__col">
+
+			<button class="e-mq-bar__settings-button e-mq-bar__button">
+				<span class="elementor-screen-only">Close</span>
+				<i class="eicon-cog" aria-hidden="true"></i>
+			</button>
+
+			<button class="e-mq-bar__close-button e-mq-bar__button">
+				<span class="elementor-screen-only">Close</span>
+				<i class="eicon-close" aria-hidden="true"></i>
```
Translate.

matipojo commented on includes/editor-templates/responsive-top-bar.php:
```diff
@@ -0,0 +1,48 @@
+		<div class="e-mq-bar__col">
+
+			<button class="e-mq-bar__settings-button e-mq-bar__button">
+				<span class="elementor-screen-only">Close</span>
```
```suggestion
				<span class="elementor-screen-only">Settings</span>
```

matipojo commented on assets/dev/js/editor/editor-base.js:
```diff
@@ -14,6 +14,7 @@ import Navigator from './regions/navigator/navigator';
 import NoticeBar from './utils/notice-bar';
 import Preview from 'elementor-views/preview';
 import PopoverToggleControl from 'elementor-controls/popover-toggle';
+import ResponsiveBar from './regions/responsive-bar/responsive-bar';
```
A. Is this an experiment?
B. Why not work with commands?
C. Screenshots before/after will be nice.

joshmarom replied:
@kobizz 

Any comments?

kobizz replied:
A. No
B. Will do in the next phase
C. @joshmarom please attach here relevant screenshots

joshmarom replied:
![Screen Shot 2021-02-14 at 21 02 16](https://user-images.githubusercontent.com/2726148/107945170-e9e86b00-6f97-11eb-8c80-24d8543bffd3.png)
![Screen Shot 2021-02-14 at 21 03 23](https://user-images.githubusercontent.com/2726148/107945199-f5d42d00-6f97-11eb-80c6-4848b7e35d14.png)
![Screen Shot 2021-02-14 at 21 03 10](https://user-images.githubusercontent.com/2726148/107945219-fa004a80-6f97-11eb-91f1-8bc5f5d4d0f3.png)

joshmarom committed:
Use translation strings and tidy code

joshmarom committed:
Refactor component name to `responsive-bar`
